politeiawww/user/mysql/mysql.go

```diff
@@ -0,0 +1,1028 @@
+		if err != nil {
+			return fmt.Errorf("encrypt user '%v': %v",
+				v.ID, err)
+		}
+
+		v.Blob = eb
+		// Store new user blob.
+		_, err = tx.ExecContext(ctx,
+			"UPDATE users SET uBlob = ? WHERE ID = $1", v.Blob, v.ID)
+		if err != nil {
+			return fmt.Errorf("save user '%v': %v", v.ID, err)
+		}
+	}
+
+	// Rotate keys for sessions table.
+	type Session struct {
+		Key  string
+		Blob []byte // Encrypted blob of session data.
+	}
+	var sessions []Session
+	rows, err = tx.QueryContext(ctx, "SELECT k, sBlob FROM sessions")
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var s Session
+		if err := rows.Scan(&s.Key, &s.Blob); err != nil {
+			return err
+		}
+		sessions = append(sessions, s)
+	}
+	// Rows.Err will report the last error encountered by Rows.Scan.
+	if err := rows.Err(); err != nil {
+		return err
+	}
+
+	for _, v := range sessions {
+		b, _, err := sbox.Decrypt(oldKey, v.Blob)
+		if err != nil {
+			return fmt.Errorf("decrypt session '%v': %v",
+				v.Key, err)
+		}
+
+		eb, err := sbox.Encrypt(user.VersionSession, newKey, b)
+		if err != nil {
+			return fmt.Errorf("encrypt session '%v': %v",
+				v.Key, err)
+		}
+
+		v.Blob = eb
+		// Store new user blob.
+		_, err = tx.ExecContext(ctx,
+			"UPDATE sessions SET sBlob = $1 WHERE k = $2", v.Blob, v.Key)
+		if err != nil {
+			return fmt.Errorf("save session '%v': %v", v.Key, err)
+		}
+	}
+
+	return nil
+}
+
+// InsertUser inserts a user record into the database. The record must be a
+// complete user record and the user must not already exist. This function is
+// intended to be used for migrations between databases.
+//
+// InsertUser satisfies the Database interface.
+func (m *mysql) InsertUser(u user.User) error {
+	log.Tracef("UserInsert: %v", u.Username)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	ub, err := user.EncodeUser(u)
+	if err != nil {
+		return err
+	}
+
+	eb, err := m.encrypt(user.VersionUser, ub)
+	if err != nil {
+		return err
+	}
+
+	// Insert new user into database.
+	ur := struct {
+		ID        string
+		Username  string
+		Blob      []byte
+		CreatedAt int64
+	}{
+		ID:        u.ID.String(),
+		Username:  u.Username,
+		Blob:      eb,
+		CreatedAt: time.Now().Unix(),
+	}
+	_, err = m.userDB.ExecContext(ctx,
+		"INSERT INTO users (ID, username, uBlob, createdAt) VALUES ($1, $2, $3, $4)",
+		ur.ID, ur.Username, ur.Blob, ur.CreatedAt)
+	if err != nil {
+		return fmt.Errorf("insert user: %v", err)
+	}
+
+	return nil
+}
+
+// UserNew creates a new user record in the database.
+func (m *mysql) UserNew(u user.User) error {
+	log.Tracef("UserNew: %v", u.Username)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	// Start transaction.
+	opts := &sql.TxOptions{
+		Isolation: sql.LevelDefault,
+	}
+	tx, err := m.userDB.BeginTx(ctx, opts)
+	defer tx.Rollback()
```

Comment:
This isn't right.  Once the tx has been comitted, calls to `Rollback()` will error.  Defering this means that you're causing an error on every invocation of `UserNew`, even though you're not logging it.  The way you're using this is how the context is suppose to be used.  From the `BeginTx` comment:

```
// The provided context is used until the transaction is committed or rolled back.
// If the context is canceled, the sql package will roll back
// the transaction. Tx.Commit will return an error if the context provided to
// BeginTx is canceled.
```

The context handles rolling back the tx on expected error prior to calls to `Commit()` and `Rollback()`.  Here's how I use `Rollback()` in my current tstore tx PR.


```
	err = tx.Commit()
	if err != nil {
		if err2 := tx.Rollback(); err2 != nil {
			// We're in trouble!
			panic(fmt.Errorf("rollback tx failed: commit:'%v' rollback:'%v'",
				err, err2))
		}
		return nil, fmt.Errorf("commit tx: %v", err)
	}
   
```

Apply to all calls in the PR.

Reply:
hrmm i think i was trying to follow this one here:
https://stackoverflow.com/questions/46421602/why-defer-a-rollback
 
 with this answer here:
 https://stackoverflow.com/a/62533516/8148239

Reply:
Context support is a relatively new addition to the sql package.  They're not using a context in that example, so deferring the rollback is the best way to do it.  When you use the sql API that allows for contexts, use the way I outlined.

Reply:
Oh I see, makes sense, will update.